sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c

```diff
@@ -213,6 +339,218 @@ void gnrc_ipv6_nib_rtr_adv_pio_cb(gnrc_netif_t *upstream, const ndp_opt_pi_t *pi
+ *            was already processed
+ * @return -1 if the sender l2 address is in order behind the local l2 address
+ */
+static int _get_my_l2addr_rank(gnrc_netif_t *iface, gnrc_pktsnip_t *pkt)
+{
+    const void *src_addr;
+    gnrc_pktsnip_t *netif_hdr;
+    gnrc_netif_hdr_t *hdr;
+
+    if (iface == NULL) {
+        return 0;
+    }
+
+    netif_hdr = gnrc_pktsnip_search_type(pkt, GNRC_NETTYPE_NETIF);
+    if (netif_hdr == NULL) {
+        return 0;
+    }
+
+    /* ignore packet if it was received on the wrong interface */
+    hdr = netif_hdr->data;
+    if (iface->pid != hdr->if_pid) {
+        return 0;
+    }
+
+    /* ignore packets without source address */
+    src_addr = gnrc_netif_hdr_get_src_addr(hdr);
+    if (src_addr == NULL) {
+        return 0;
+    }
+
+    /* check if we have seen the host before */
+    if (_alloc_l2addr_entry(src_addr, iface->l2addr_len) == 0) {
+        return 0;
+    }
+
+    return memcmp(iface->l2addr, src_addr, iface->l2addr_len);
+}
+
+static void *_eventloop(void *arg)
+{
+    (void)arg;
+
+    xtimer_t timeout_timer;
+    msg_t msg, timeout_msg = { .type = SERVER_MSG_TYPE_TIMEOUT };
+    gnrc_netreg_entry_t server = GNRC_NETREG_ENTRY_INIT_PID(0, KERNEL_PID_UNDEF);
+    uint8_t idx_start = 0;
+    uint8_t subnets = gnrc_netif_numof() - 1;
+    uint8_t resend_tx = CONFIG_GNRC_IPV6_AUTO_SUBNETS_TX_RETRIES;
+
+    DEBUG("auto_subnets: %u local subnets\n", subnets);
+
+    if (subnets == 0) {
+        return NULL;
+    }
+
+    /* setup the message queue */
+    msg_init_queue(server_queue, SERVER_MSG_QUEUE_SIZE);
+
+    /* register server to receive messages from given port */
+    gnrc_netreg_entry_init_pid(&server, CONFIG_GNRC_IPV6_AUTO_SUBNETS_PORT, thread_getpid());
+    gnrc_netreg_register(GNRC_NETTYPE_UDP, &server);
+
+    while (1) {
+        msg_receive(&msg);
+
+        switch (msg.type) {
+        case GNRC_NETAPI_MSG_TYPE_RCV:
+        {
+            gnrc_pktsnip_t *pkt = msg.content.ptr;
+            auto_subnets_request_t *request = pkt->data;
+
+            /* check if we already got an announcement from that host */
+            int res = _get_my_l2addr_rank(_upstream, pkt);
+            if (res) {
+                /* calculate total number of subnets */
+                subnets += request->num_subnets;
+
+                DEBUG("auto_subnets: %u new remote subnets, total %u\n",
+                      request->num_subnets, subnets);
+
+                /* If other host is before us in order of MAC address, add
+                 * their subnets to our offset */
+                if (res > 0) {
+                    idx_start += request->num_subnets;
+                }
+            }
+
+            gnrc_pktbuf_release(pkt);
+            break;
+        }
+        case SERVER_MSG_TYPE_TIMEOUT:
+            if (resend_tx--) {
+                uint32_t timeout_us;
+                auto_subnets_request_t request = {
+                    .num_subnets = gnrc_netif_numof() - 1,
```

Comment:
The variable `subnets` gets the same value on initialization above. While it is incremented by received packets, its initial value could be used to set another variable, that is used here instead of counting the network interfaces every time.

Reply:
since this is not time critical (only gets called once every time a RA is received) I thought there was little reason to spend memory on caching the result.

Reply:
Welp, might also save some ROM ;-).